### PR TITLE
Copy files from one location to another

### PR DESCRIPTION
**Pull request by etorreborre**
And use the native "copy" when from/to are on the same file system.

### CODE REVIEW

File: notion-core/src/main/scala/com/ambiata/notion/core/LocationIO.scala [hunk 1]
```diff
@@ -144,6 +144,32 @@ case class LocationIO(configuration: Configuration, s3Client: AmazonS3Client) {
...
+      _      <-
+      if (exists)
+        if (overwrite) delete(to) >> pipe(from, to)
+        else           RIO.fail(s"There is already a file at ${to.render}")
+      else pipe(from, to)
+    } yield ()
+  }
+
+  /** pipe bytes from one location to another */
+  def pipe(from: Location, to: Location): RIO[Unit] =
```
Comment (markhibberd) [inline notion-core/src/main/scala/com/ambiata/notion/core/LocationIO.scala]: /minor Just the name/comment here doesn't make a lot of sense here any more, it is basically just unsafeCopyFile (unsafe just because it doesn't do the override here. I would consider inlining, but if not, I would rename to something that fits a bit better.

File: notion-core/src/test/scala/com/ambiata/notion/core/LocationIOSpec.scala [hunk 2]
```diff
@@ -120,6 +121,17 @@ class LocationIOSpec extends Specification with ScalaCheck { def is = s2"""
     e <- i.exists(p)
   } yield e ==== true)
 
+  def copy = prop((from: LocationTemporary, to: LocationTemporary, lines: List10[N]) =>
+    for {
+      f  <- from.location
+      i  <- from.io
+      _  <- i.writeUtf8Lines(f, lines.value.map(_.value))
```
Comment (markhibberd) [inline notion-core/src/test/scala/com/ambiata/notion/core/LocationIOSpec.scala]: Could you either change this to be a blob of raw bytes or add another test. This limits the test a bit more than it should I think.

**Comment by markhibberd** [comment]
:+1: But out of curiosity, what drove this change, I wouldn't expect this to help on local disk - on hdfs it might if it does less shuffling server side, but I don't know if it does and would be somewhat surprised.

**Comment by etorreborre** [comment]
Piping S3 to S3 was using the /tmp dir. That's why I did this change.

File: notion-core/src/test/scala/com/ambiata/notion/core/LocationIOSpec.scala [hunk 3]
```diff
@@ -120,6 +122,18 @@ class LocationIOSpec extends Specification with ScalaCheck { def is = s2"""
     e <- i.exists(p)
   } yield e ==== true)
 
+  def copy = prop((from: LocationTemporary, to: LocationTemporary, bytes: Array[Byte]) =>
+    for {
+      f   <- from.location
+      i   <- from.io
+      s1   = new String(bytes, "UTF-8")
```
Comment (charleso) [inline notion-core/src/test/scala/com/ambiata/notion/core/LocationIOSpec.scala]: Curious why you didn't use a generated String directly?

Comment (markhibberd) [inline notion-core/src/test/scala/com/ambiata/notion/core/LocationIOSpec.scala]: This is my fault, I confused the situation with a previous comment, but this isn't what I meant. This is going to break (well at least the other one that test the symmetric nature, random bytes encoded to strings aren't symmetric). What I was getting at with previous comment is that strings are irrelevant for copying a file, we really need to get out of the particularly dangerous habit of believing all content is a string - we need to be checking that copy is an exact byte-for-byte copy, there should be no strings in any of these tests, at best it is irrelevant and at worst it hides very dangerous corner cases that will lead to corruption.

Comment (charleso) [inline notion-core/src/test/scala/com/ambiata/notion/core/LocationIOSpec.scala]: > we need to be checking that copy is an exact byte-for-byte copy

Agreed.
